### PR TITLE
[SofaMiscForceField] Fix addForce function in MeshMatrixMass

### LINKED ISSUE #2301: MeshMatrixMass : lumped vs sparse incoherent
According to this [PR on Caribou's repo](https://github.com/jnbrunet/caribou/pull/60), the MeshMatrixMass::addForce isn't giving the same result either your are using a lumped scheme or not, and I'm not talking about a small difference. You can test it against SOFA v20.12, which works correctly.

This may be a critical issue.

### PR DESCRIPTION
**Pull request by hugtalbot**
Fix bug introduced in #2193 and reported by @jnbunet in the Caribou CI! Thanks again @jnbrunet and @epernod for the discussions !

- Homogeneize the use of the massLumpingCoeff
- In the test, only cosmetic change
- homogeneize and add scenes in the regression

Fixes #2301

[ci-depends-on https://github.com/sofa-framework/Regression/pull/18]

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 1]
```diff
@@ -107,8 +107,13 @@ template< class DataTypes, class MassType>
...
     SOFA_UNUSED(id);
-    helper::WriteAccessor<Data<Real> > totalMass(this->m->d_totalMass);
-    totalMass -= EdgeMass;
+    if(!this->m->isLumped())
+    {
+        helper::WriteAccessor<Data<Real> > totalMass(this->m->d_totalMass);
+        totalMass -= EdgeMass;
+    }
+    else
+        SOFA_UNUSED(EdgeMass);
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
    else
    {
        SOFA_UNUSED(EdgeMass);
    }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 2]
```diff
@@ -166,7 +171,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyTriangleCreati
                 VertexMasses[ t[j] ] += mass;
 
             // update total mass: 
-            totalMass += 3.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass += 3.0 * mass;
+            }
+            else
+                totalMass += 3.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass += 3.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 3]
```diff
@@ -261,7 +274,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyTriangleDestru
                 VertexMasses[ t[j] ] -= mass;
 
             // update total mass
-            totalMass -= 3.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass -= 3.0 * mass;
+            }
+            else
+                totalMass -= 3.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass -= 3.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 4]
```diff
@@ -401,7 +422,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyQuadCreation(c
                 VertexMasses[ q[j] ] += mass;
 
             // update total mass
-            totalMass += 4.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass += 4.0 * mass;
+            }
+            else
+                totalMass += 4.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass += 4.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 5]
```diff
@@ -496,7 +525,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyQuadDestructio
                 VertexMasses[ q[j] ] -= mass;
 
             // update total mass
-            totalMass -= 4.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass -= 4.0 * mass;
+            }
+            else
+                totalMass -= 4.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass -= 4.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 6]
```diff
@@ -638,7 +675,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyTetrahedronCre
                 VertexMasses[ t[j] ] += mass;
 
             // update total mass
-            totalMass += 4.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass += 4.0 * mass;
+            }
+            else
+                totalMass += 4.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass += 4.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 7]
```diff
@@ -733,7 +778,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyTetrahedronDes
                 VertexMasses[ t[j] ] -= mass;
 
             // update total mass
-            totalMass -= 4.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass -= 4.0 * mass;
+            }
+            else
+                totalMass -= 4.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass -= 4.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 8]
```diff
@@ -874,7 +927,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyHexahedronCrea
                 VertexMasses[ h[j] ] += mass;
 
             // update total mass
-            totalMass += 8.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass += 8.0 * mass;
+            }
+            else
+                totalMass += 8.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass += 8.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl [hunk 9]
```diff
@@ -969,7 +1030,12 @@ void MeshMatrixMass<DataTypes, MassType>::VertexMassHandler::applyHexahedronDest
                 VertexMasses[ h[j] ] -= mass;
 
             // update total mass
-            totalMass -= 8.0 * mass * MMM->m_massLumpingCoeff;
+            if(!this->m->isLumped())
+            {
+                totalMass -= 8.0 * mass;
+            }
+            else
+                totalMass -= 8.0 * mass * MMM->m_massLumpingCoeff;
```
Comment (guparan) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
            else
            {
                totalMass -= 8.0 * mass * MMM->m_massLumpingCoeff;
            }
```

File: modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl (hunk 1 above)
Comment (alxbilger) [inline modules/SofaMiscForceField/src/SofaMiscForceField/MeshMatrixMass.inl]:
```suggestion
```

**Review by guparan — CHANGES_REQUESTED** [review]
:wave:

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

**Comment by guparan** [comment]
A lot of regression tests are failing. Maybe their references should be updated?
@hugtalbot

**Pull request merged**